serialize broadcast topology updates and coalesce updates

murali-reddy opened the pull request:
This PR is an attempt to address high topology updates gossip's that can result in certain use-cases.

- minimizes the concurrent topology gossip broadcast updates
- coalesce the topology updates when possible before sending gossip broadcast

Fixes #116

bboreham commented on local_peer.go:
```diff
@@ -147,10 +158,32 @@ func (peer *localPeer) actorLoop(actionChan <-chan localPeerAction) {
 	}
 }
 
+func (peer *localPeer) brodcastTopologyUpdates() {
```
spelling: brod -> broad

bboreham commented on local_peer.go:
```diff
@@ -147,10 +158,32 @@ func (peer *localPeer) actorLoop(actionChan <-chan localPeerAction) {
+			peer.brodcastTopologyUpdates()
+			if !peer.timer.Stop() {
+				select {
+				case <-peer.timer.C:
```
we don't need this in the case we've already read from `C`

bboreham commented on local_peer.go:
```diff
@@ -244,7 +272,13 @@ func (peer *localPeer) broadcastPeerUpdate(peers ...*Peer) {
+			peer.timer.Reset(topologyUpdateInterval)
+			peer.pendingTopologyUpdate = true
+		}
+		peer.topologyUpdatesChan <- append(peers, peer.Peer)
```
suggest accumulating in a slice here because `chan` is fixed-size so will block

bboreham commented on local_peer.go:
```diff
@@ -8,13 +8,20 @@ import (
 )
 
+const (
+	topologyUpdateInterval = 1 * time.Second
```
"update interval" isn't quite how I think about this.
What about "defer duration" ?

murali-reddy commented:
updated PR as per review comments. PTAL

bboreham commented on router.go:
```diff
@@ -229,14 +229,19 @@ func (router *Router) sendPendingGossip() bool {
+		for _, p := range update {
+			names[p.Name] = struct{}{}
+		}
+		gossipData = gossipData.Merge(&topologyGossipData{peers: router.Peers, update: names})
```
This seems unnecessary - all this `Merge` does is union the two maps, so we can just build up one set then send that at the end.

bboreham commented on local_peer.go:
```diff
@@ -8,13 +8,20 @@ import (
-	actionChan chan<- localPeerAction
+	router                *Router
+	actionChan            chan<- localPeerAction
+	topologyUpdates       [][]*Peer
```
Given that we're eventually going to build a set of peer names from this data, why not hold a `peerNameSet` here?

bboreham commented on local_peer.go:
```diff
@@ -147,10 +158,25 @@ func (peer *localPeer) actorLoop(actionChan <-chan localPeerAction) {
 	}
 }
 
+func (peer *localPeer) broadcastTopologyUpdates() {
```
maybe `broadcastPendingTopologyUpdates` ?

bboreham commented on local_peer.go:
```diff
@@ -147,10 +158,23 @@ func (peer *localPeer) actorLoop(actionChan <-chan localPeerAction) {
 }
 
+func (peer *localPeer) broadcastPendingTopologyUpdates() {
+	peer.router.Routes.recalculate()
```
Because this is async, I think it's unwise to move it here.
I'm not entirely sure it was ever necessary, but left at the original place there is some chance it will happen before the broadcast.

murali-reddy replied:
> I'm not entirely sure it was ever necessary

I changed it so to avoid recalculate an every add/delete connection. But considering #111 
I reverted to keep original calls to routes.recalculate().

murali-reddy commented:
>Sorry I need to withdraw my approval as unit tests are failing. I couldn't see an easy fix.

Fixed the UT. Added `sendPendingTopologyUpdates` in similar lines to `sendPendingGossip`
